### DISCUSSION TITLE
Transformation that can raise an error?

### DISCUSSION THREAD

**Discussion by cakoose** [colinhacks/zod#345] [Q&A]
I have a transformation step that might also raise an error.  For example, something like this:

```ts
const fileModes = new Map([['read', 1], ['write', 2], ['execute', 4]]);

const operationSchema = z.object({
    mode: z.string().transformAndCheck(s => {
        const i = fileModes.get(s);
        if (i === undefined) return {pass: false, error: ...};
        return {pass: true, value: i};
    }),
});
```

Is there a good way to express that in Zod?

**Comment by colinhacks** [comment]
Not a bad idea...I'll need to think more about this. 

Currently you'd have to do this check as a refinement first, then transform it once you're confident that the data is valid.

```ts
z.string().refine(s => !!fileModes.get(s), { message: "Invalid file mode"  }).transform(s => {
        const i = fileModes.get(s);
        return {pass: true, value: i};
    }),
```

**Reply by cakoose** [reply]
The framework I used at my last job just had the combined operation.  It feels like a better match for the "parse don't validate" philosophy.

(Edit: But for now the check+transform thing works.  I can probably create a helper to reduce the boilerplate a bit.)

**Reply by colinhacks** [reply]
What framework was that?

**Reply by cakoose** [reply]
It was something internal.  Not open source.

But the basic idea is that everything works like `.safeParse`.

**Reply by oleg-codaio** [reply]
Thoughts on something like this? https://github.com/colinhacks/zod/pull/420

**Reply by cakoose** [reply]
@oleg-codaio: How would the original example look with `.refineWith`?

**Reply by oleg-codaio** [reply]
I don't think this example actually needs `refineWith`, because you could write it like this currently:

```ts
const fileModes = new Map([['read', 1], ['write', 2], ['execute', 4]]);

const operationSchema = z.object({
    mode: z.string().enum(Object.keys(fileModes)).transform(m => fileModes.get(m)),
});
```

I guess if you were to, you could write:

```ts
const operationSchema = z.object({
    mode: z.string().transform(m => fileModes.get(m)).refineWith(z.enum(Object.values(fileModes))),
});

```

**Reply by exrhizo** [reply]
Hm, I am running into a similar situation, I implemented this:
https://github.com/coordinape/coordinape/pull/368

Could we have this interface?

It seems it was a design decision not to allow errors, but sometimes when I am interfacing with other code it becomes desirable.

**Reply by exrhizo** [reply]
@colinhacks happy new year!
Any thoughts about me giving it a try to create in the project?
idk, if it comes up in our project again, I'd like to be able to resolve
